File: ExtremeParser/Parsers/IfDefines.vb
Imports System.Text.RegularExpressions

Public Class IfDefines
    Public Shared Sub Parse(ByRef code As String, filePath As String, prjPath As String, ByRef parts As CodeParts, Optional add As Boolean = True)
        For Each match As Match In Regex.Matches(code, "\#if[ \t]+(!)?defined[ \t]+(.+)([\s\S]*?)\#endif", RegexOptions.Multiline)
            Dim isNt As Boolean = IIf(match.Groups(1).Value = "", False, True)
            Dim condition As String = ""
            Dim mainCode As String = ""
            Dim elseClode As String = " "

            'Start filling the vars.
            condition = match.Groups(2).Value.Trim

            If match.Groups(3).Value.Contains("#else") Then
                Dim s As String() = Split(match.Groups(3).Value, "#else")
                mainCode = s(0).Trim : elseClode = s(1).Trim
            Else
                mainCode = match.Groups(3).Value.Trim
            End If

            'The result of the parse will be saved here for deletion.
            Dim result As IfDefinedParser = Nothing

            'Now check which part needs to be parsed by seeing isNt and the else.
            If isNt = False Then
                'Here the thing should BE defined
                If isDefined(parts, condition) = False Then
                    'Parse the main.
                    result = New IfDefinedParser(parts, mainCode, filePath, prjPath, False)
                Else
                    'Parse the else.
                    result = New IfDefinedParser(parts, elseClode, filePath, prjPath, False)
                End If
            Else
                'It should NOT be defined.
                If isDefined(parts, condition) = True Then
                    'Parse the main.
                    result = New IfDefinedParser(parts, mainCode, filePath, prjPath, False)
                Else
                    'Parse the else.
                    result = New IfDefinedParser(parts, elseClode, filePath, prjPath, False)
                End If
            End If
        Next
    End Sub

    Private Shared Function isDefined(ByRef parts As CodeParts, str As String)
        If parts.Defines.FindIndex(Function(x) x.Value.DefineName = str) <> -1 Then Return True
        If parts.Macros.FindIndex(Function(x) x.Value.DefineName = str) <> -1 Then Return True
        If parts.Enums.FindIndex(Function(x) x.Value.EnumName = str) <> -1 Then Return True
        If parts.Publics.FindIndex(Function(x) x.Value.FuncName = str) <> -1 Then Return True
        If parts.Stocks.FindIndex(Function(x) x.Value.FuncName = str) <> -1 Then Return True
        If parts.Functions.FindIndex(Function(x) x.Value.FuncName = str) <> -1 Then Return True
        If parts.Natives.FindIndex(Function(x) x.Value.FuncName = str) <> -1 Then Return True
        If parts.publicVariables.FindIndex(Function(x) x.Value.VarName = str) <> -1 Then Return True

        Return False
    End Function
End Class

File: ExtremeParser/Parsers/IfDefines.cs
using System;
using System.Text.RegularExpressions;

namespace ExtremeParser.Parsers
{
    public class IfDefines
    {
        public static void Parse(string code, string filePath, string prjPath, ref CodeParts parts, bool add = true)
        {
            foreach (Match match in Regex.Matches(code, "\\#if[ \\t]+(!)?defined[ \\t]+(.+)([\\s\\S]*?)\\#endif",
                RegexOptions.Multiline))
            {
                bool isNt = (match.Groups[1].Value != "");
                string condition = "";
                string mainCode = "";
                string elseClode = " ";

                //Start filling the vars.
                condition = Convert.ToString(match.Groups[2].Value.Trim());

                if (match.Groups[3].Value.Contains("#else"))
                {
                    string[] s = Convert.ToString(match.Groups[3].Value).Split(new [] {"#else"}, StringSplitOptions.None);
                    mainCode = s[0].Trim();
                    elseClode = s[1].Trim();
                }
                else
                {
                    mainCode = Convert.ToString(match.Groups[3].Value.Trim());
                }

                //The result of the parse will be saved here for deletion.
                Parser result = null;

                //Now check which part needs to be parsed by seeing isNt and the else.
                if (isNt == false)
                {
                    //Here the thing should BE defined
                    if (IsDefined(parts, condition) == false)
                    {
                        //Parse the main.
                        result = new Parser(parts, mainCode, filePath, prjPath, false, true);
                    }
                    else
                    {
                        //Parse the else.
                        result = new Parser(parts, elseClode, filePath, prjPath, false, true);
                    }
                }
                else
                {
                    //It should NOT be defined.
                    if (IsDefined(parts, condition) == true)
                    {
                        //Parse the main.
                        result = new Parser(parts, mainCode, filePath, prjPath, false, true);
                    }
                    else
                    {
                        //Parse the else.
                        result = new Parser(parts, elseClode, filePath, prjPath, false, true);
                    }
                }
            }
        }

        private static dynamic IsDefined(CodeParts parts, string str)
        {
            if (parts.Defines.FindIndex(x => x.DefineName == str) != -1)
            {
                return true;
            }

            if (parts.Macros.FindIndex(x => x.DefineName == str) != -1)
            {
                return true;
            }

            if (parts.Enums.FindIndex(x => x.EnumName == str) != -1)
            {
                return true;
            }

            if (parts.Publics.FindIndex(x => x.FuncName == str) != -1)
            {
                return true;
            }

            if (parts.Stocks.FindIndex(x => x.FuncName == str) != -1)
            {
                return true;
            }

            if (parts.Functions.FindIndex(x => x.FuncName == str) != -1)
            {
                return true;
            }

            if (parts.Natives.FindIndex(x => x.FuncName == str) != -1)
            {
                return true;
            }

            if (parts.PublicVariables.FindIndex(x => x.VarName == str) != -1)
            {
                return true;
            }

            return false;
        }
    }
}